src/components/PasswordToggle/PasswordToggle.tsx

```diff
@@ -0,0 +1,135 @@
+import classNames from "classnames";
+import React, { useEffect, useRef, useState } from "react";
+import type { InputHTMLAttributes, ReactNode } from "react";
+
+import Button from "../Button";
+import Field from "../Field";
+import Label from "../Label";
+
+import type { ClassName, PropsWithSpread } from "types";
+
+/**
+ * The props for the Input component.
+ */
+export type Props = PropsWithSpread<
+  {
+    /**
+     * The content for caution validation.
+     */
+    caution?: ReactNode;
+    /**
+     * Optional class(es) to pass to the input element.
+     */
+    className?: ClassName;
+    /**
+     * The content for error validation.
+     */
+    error?: ReactNode;
+    /**
+     * Help text to show below the field.
+     */
+    help?: ReactNode;
+    /**
+     * The id of the input.
+     */
+    id: string;
+    /**
+     * The label for the field.
+     */
+    label?: ReactNode;
+    /**
+     * Whether the field is read only.
+     */
+    readOnly?: boolean;
+    /**
+     * Whether the field is required.
+     */
+    required?: boolean;
+    /**
+     * The content for success validation.
+     */
+    success?: ReactNode;
+    /**
+     * Whether to focus on the input on initial render.
+     */
+    takeFocus?: boolean;
```

Comment:
I'm not sure about this... 

Supporting a `forwardRef` might be more appropriate. It would enable this functionality by using a more standard `inputRef.focus()` from the parent as well as any other ref-based interactions.

Reply:
I believe they based it on existing `Input` component to keep the props consistent: https://canonical-web-and-design.github.io/react-components/?path=/docs/input--text-input

If we want to change the approach to use of refs we should do it in both places I guess.

react-components repo is kind of museum of how our React experience evolved, so when adding new stuff it's always a compromise between being consistent with things that are already there and using newer/better approaches 😄

Reply:
@hatched We've just updated it, and it works, but we don't _really_ understand why - when you're around later we'd love a masterclass!